In [52]:
import pandas as pd

df=pd.read_csv('sales.csv')
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,43,CA-2016-101343,17-Jul-16,22-Jul-16,Standard Class,RA-19885,Ruben Ausman,Corporate,United States,Los Angeles,...,90049,West,OFF-ST-10003479,Office Supplies,Storage,"Eldon Base for stackable storage shelf, platinum",77.88,2,0.0,3.8940
1,514,CA-2017-163405,21-Dec-17,25-Dec-17,Standard Class,BN-11515,Bradley Nguyen,Consumer,United States,Los Angeles,...,90049,West,OFF-AR-10003811,Office Supplies,Art,Newell 327,6.63,3,0.0,1.7901
2,515,CA-2017-163405,21-Dec-17,25-Dec-17,Standard Class,BN-11515,Bradley Nguyen,Consumer,United States,Los Angeles,...,90049,West,OFF-AR-10001246,Office Supplies,Art,Newell 317,5.88,2,0.0,1.7052
3,1606,US-2016-115819,19-Apr-16,24-Apr-16,Second Class,JO-15280,Jas O'Carroll,Consumer,United States,Los Angeles,...,90049,West,OFF-AR-10000823,Office Supplies,Art,Newell 307,5.46,3,0.0,1.5288
4,1607,US-2016-115819,19-Apr-16,24-Apr-16,Second Class,JO-15280,Jas O'Carroll,Consumer,United States,Los Angeles,...,90049,West,OFF-AR-10004456,Office Supplies,Art,Panasonic KP-4ABK Battery-Operated Pencil Shar...,73.20,5,0.0,21.2280


In [53]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   str    
 2   Order Date     9994 non-null   str    
 3   Ship Date      9994 non-null   str    
 4   Ship Mode      9994 non-null   str    
 5   Customer ID    9994 non-null   str    
 6   Customer Name  9994 non-null   str    
 7   Segment        9994 non-null   str    
 8   Country        9994 non-null   str    
 9   City           9994 non-null   str    
 10  State          9994 non-null   str    
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   str    
 13  Product ID     9994 non-null   str    
 14  Category       9994 non-null   str    
 15  Sub-Category   9994 non-null   str    
 16  Product Name   9994 non-null   str    
 17  Sales          9994 non-null   float64
 18  Quantity       9994

In [54]:
df.describe()

,Row ID,Postal Code,Sales,Quantity,Discount,Profit
count,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000
mean,4997.500000,55190.379428,229.858001,3.789574,0.156203,28.656896
std,2885.163629,32063.693350,623.245101,2.225110,0.206452,234.260108
min,1.000000,1040.000000,0.444000,1.000000,0.000000,-6599.978000
25%,2499.250000,23223.000000,17.280000,2.000000,0.000000,1.728750
50%,4997.500000,56430.500000,54.490000,3.000000,0.200000,8.666500
75%,7495.750000,90008.000000,209.940000,5.000000,0.200000,29.364000
max,9994.000000,99301.000000,22638.480000,14.000000,0.800000,8399.976000


In [55]:
df.isnull().sum()

Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64

When it comes to handling missing values or null values the procedure is this ...

1.You use the .info() to get a view of all the rows, columns , not null values and the datatypes
  Now the not null will enable you to know the number of null values , the data type will enable you to narrow down to the exact thing you will use to fill the null values
  If the data type is categorical data(words) you use mode
  If the data type is numerical values then you can use either median or mode 
  To determine if its either median or mode you are going to be using ...there are some certain aspect you should consider ie the distribution of the data in the column

  (The reason why we are plotting this data is because by doing this you will be answering the question By how much is the data spread)

  You have to plot that column using a histogram and ensure that the frequency polygon is turned on i.e (kde=True)
  You also have to confirm the distribution using a boxplot so that you know if the distribution you're getting is the actual one or not

  You can use the .skew() so that you can know the skewness of the data.ie df['Age'].skew()

  INTERPRETATION:
  If you get a positive number it means its right skewed 
  If you get a negative number it means its left skewed 



  SUMMARY :
  WE ARE CURRENTLY IN DATA PREPROCESSING BIT OF OUR DATA SCIENCE LIFE CYCLE
    1.We inspect the data using the .info() and the .describe()
    2.The .info():Gives us the rows, columns , not null count and the data type
    3.Since our data didn't contain any null values we continued to inspect the data by checking the outliers ...now these



In [56]:
q1=df['Sales'].quantile(0.25)
q3=df['Sales'].quantile(0.75)
iqr=q3-q1
lower_bound=q1-1.5*iqr
upper_bound=q3+1.5*iqr


In [ ]:
# Generating the outliers
upper_outlier=df[
    (df['Sales']>upper_bound)
]

lower_outlier=df[
    (df['Sales'] < lower_bound)
]

We wanna inspect the the sales that are above the normal range i.e the upper outliers.
To do this I have to answer a couple of questions such as which Region contains the most sales in the upper outlier which is what i'm doing at the moment 


In [ ]:
upper_outlier['Region'].value_counts()


Region
West       374
East       334
Central    257
South      202
Name: count, dtype: int64

WE CAN ALSO CHECK THE PRODUCT IN THOSE REGIONS 

In [102]:
products_in_upper=upper_outlier[['Product ID','Product Name']]


Next Question to answer is which customer frequently makes these types of purchases


In [70]:
upper_outlier[['Customer ID','Customer Name']].value_counts()

Customer ID  Customer Name   
BM-11650     Brian Moss          7
NF-18385     Natalie Fritzler    6
JE-15715     Joe Elijah          6
CL-12565     Clay Ludtke         6
ME-17320     Maria Etezadi       6
                                ..
AR-10345     Alex Russell        1
AG-10525     Andy Gerbode        1
AG-10675     Anna Gayman         1
MC-17605     Matt Connell        1
MC-18100     Mick Crebagga       1
Name: count, Length: 582, dtype: int64

Next question to answer is: What Category of Products are involved in these sales to make them outliers 


In [69]:
upper_outlier['Category'].value_counts()

Category
Furniture          467
Technology         400
Office Supplies    300
Name: count, dtype: int64

What quantities are being bought to generate the outliers in these sales ...Remember this is from a general perspective 


In [72]:
upper_outlier['Quantity'].describe()

count    1167.000000
mean        5.158526
std         2.447331
min         1.000000
25%         3.000000
50%         5.000000
75%         7.000000
max        14.000000
Name: Quantity, dtype: float64

Things to note from the above statistical summary 
1. The count refers to the number of rows and not the number of items sold
2. Mean refers to the average number items sold in the upper_outlier


LETS PROCEED TO DISCOUNTS 

The question I wanna answer is if I were to compare my outlier's discount mean to that of the whole dataset will they differ or are they the same 


In [74]:
df['Discount'].mean()

np.float64(0.15620272163297977)

In [76]:
upper_outlier['Discount'].mean()

np.float64(0.14558697514995714)

DISCOVERY 
The average discounts in the whole dataset is 15.62 percent while the average discount in the upper_outlier is 14.56 .The difference is slightly small i.e 1.06


Although I want to see if there are other differences between the outlier and the whole dataset so i'll be compairing the statistical summary of the said 2 

In [78]:
upper_outlier['Discount'].describe()

count    1167.000000
mean        0.145587
std         0.162669
min         0.000000
25%         0.000000
50%         0.200000
75%         0.200000
max         0.800000
Name: Discount, dtype: float64

In [79]:
df['Discount'].describe()

count    9994.000000
mean        0.156203
std         0.206452
min         0.000000
25%         0.000000
50%         0.200000
75%         0.200000
max         0.800000
Name: Discount, dtype: float64

DISCOVERY :
The Standard deviation of the discounts of the whole dataset is higher than that of the outlier.What this basically means in lame man language is that the discounts in the outlier have a more consistent value as compared to the whole dataset .

IN SUMMARY :
High-sales outliers do not appear to receive substantially different discounts from the overall dataset. Their median discount is the same, their mean differs only slightly, and their discount values show somewhat less variability."

Next question is: Are these high sales values generating the company a profit or not ?


WE USE THE STATISTICAL SUMMARY :
1. The count basically means how many transactions generated a profit
2. The mean basically means the average profit we generate from each transaction
3. The standard deviation means the deviation in terms of profit margin
4. The min basically means the worst case profit;in other words it means the highest loss
5. 25% means the lower quantile profit values
6. 50% means the Median Profit
7. 75% means the upper quantile profit value
8. Max means the highest Profit Made 

In [82]:
upper_outlier['Profit'].describe()

count    1167.000000
mean      160.708308
std       657.523657
min     -6599.978000
25%        10.075550
50%       119.686000
75%       270.363600
max      8399.976000
Name: Profit, dtype: float64

In [83]:
df['Profit'].describe()

count    9994.000000
mean       28.656896
std       234.260108
min     -6599.978000
25%         1.728750
50%         8.666500
75%        29.364000
max      8399.976000
Name: Profit, dtype: float64

Discovery is that the mean profit in the high sales value is more compared to that of the whole dataset ;from a business side of things it seems like we are generating more profit when it comes to the high sales values as compared to that of the normal transactions

The profitability of high-sales transactions is more unpredictable/less consistent than profitability across normal transactions;this is brought about by the difference in standard deviation when it comes to the standard deviation in the outlier and that of the whole dataset

The highest loss was also generated from these high sales i.e min -6599.978000
The highest profit was also generated from these high sales i.e max 8399.976000


POINT TO NOTE:The quantiles are not used to generate the conclusion that we are making larger profit , they are used to show where the individual transactions sit when it comes to the profit table 




Now that we have discovered that high sales value produce a higher profit than normal transactions but we have a higher variation in profit as compared to the normal transactions ...The next question becomes why... 


In [105]:
losses=upper_outlier[upper_outlier['Profit']< 0]
losses.shape

(259, 21)

DISCOVERY:We have realised that 259 high value transactions resulted in losses

In [107]:
percentage_of_loss=(len(losses)/len(upper_outlier))*100
percentage_of_loss

22.193658954584407

DISCOVERY: The Percentage of loss as per the High Sales Value Transactions is 22.19
What this means is that even though we are generating profits in our high sales value transactions, 22.19 percent of those transactions result in losses 

With the above discovery in hand , the next question becomes now that we've realised we are making these losses , what product is most likely generating the most losses 


In [118]:
losses.groupby('Product Name')['Profit'].sum().sort_values()

Product Name
Cubify CubeX 3D Printer Double Head Print                           -9239.9692
GBC DocuBind P400 Electric Binding System                           -6859.3896
Lexmark MX611dhe Monochrome Laser Printer                           -5269.9690
GBC Ibimaster 500 Manual ProClick Binding System                    -4794.1740
GBC DocuBind TL300 Electric Binding System                          -4162.0336
                                                                       ...    
Logitech P710e Mobile Speakerphone                                    -15.4494
Fellowes Super Stor/Drawer Files                                      -14.5395
Deflect-o RollaMat Studded, Beveled Mat for Medium Pile Carpeting     -12.9122
Global Manager's Adjustable Task Chair, Storm                         -12.0784
Global Adaptabilities Conference Tables                                -8.4294
Name: Profit, Length: 148, dtype: float64

Discovery: The Cubify CubeX 3D Printer (Double Head) recorded the highest loss by a significant margin, amounting to 9,239.97.

We can also answer another question i.e Which product has featured the most when it comes to the high value transactions that generate losses .Remember the above question is different from this one because , this one focuses on the number of times a product has been featured in transactions that result in losses and the one above basically means the product that has generated the highest loss 



In [119]:
losses[['Product Name']].value_counts()

Product Name                                                                             
Hon 2090 “Pillow Soft” Series Mid Back Swivel/Tilt Chairs                                    5
Tennsco Double-Tier Lockers                                                                  5
GBC Ibimaster 500 Manual ProClick Binding System                                             5
Bretford Rectangular Conference Table Tops                                                   5
Bush Advantage Collection Racetrack Conference Table                                         5
                                                                                            ..
Bush Andora Conference Table, Maple/Graphite Gray Finish                                     1
Office Star - Mid Back Dual function Ergonomic High Back Chair with 2-Way Adjustable Arms    1
Sauder Mission Library with Doors, Fruitwood Finish                                          1
Hon 2111 Invitation Series Straight Table              

Lets Analyse these products that are generating these losses on a deeper level

In [127]:
losses.groupby('Product Name').agg(
Total_Sales=('Sales','sum'),
Total_Profit=('Profit','sum'),
Average_Discount=('Discount','mean'),
Average_Quantity=('Quantity','mean')
).sort_values('Total_Profit')


,Total_Sales,Total_Profit,Average_Discount,Average_Quantity
Product Name,,,,
Cubify CubeX 3D Printer Double Head Print,6299.979,-9239.9692,0.700000,3.500000
GBC DocuBind P400 Electric Binding System,4899.564,-6859.3896,0.766667,5.333333
Lexmark MX611dhe Monochrome Laser Printer,13769.919,-5269.9690,0.500000,5.333333
GBC Ibimaster 500 Manual ProClick Binding System,6240.036,-4794.1740,0.720000,5.800000
GBC DocuBind TL300 Electric Binding System,4395.251,-4162.0336,0.725000,4.500000
...,...,...,...,...
Logitech P710e Mobile Speakerphone,1235.952,-15.4494,0.200000,3.000000
Fellowes Super Stor/Drawer Files,1163.160,-14.5395,0.200000,4.500000
"Deflect-o RollaMat Studded, Beveled Mat for Medium Pile Carpeting",516.488,-12.9122,0.200000,7.000000


Lets also investigate the Regions 

In [130]:
losses.groupby("Region").agg(
    Total_Sales=('Sales','sum'),
    Total_Profit=('Profit','sum'),
    Average_Discount=('Discount','mean'),
    Average_Quantity=('Quantity','mean')
).sort_values('Total_Profit')

,Total_Sales,Total_Profit,Average_Discount,Average_Quantity
Region,,,,
East,99469.1810,-31223.0641,0.358889,5.222222
Central,87387.8294,-26814.3529,0.354762,5.059524
South,72547.7095,-22068.5801,0.416667,5.051282
West,37996.1080,-12278.3187,0.286957,5.326087


Discovery: The EAST region generates the highest loss as compared to the other Regions 

NOW LETS ALSO INVESTIGATE WHICH PRODUCTS GENERATE THE HIGHEST PROFITS 

In [138]:
upper_outlier.groupby('Product Name').agg(
    Total_Sales=('Sales','sum'),
    Total_Profit=('Profit','sum'),
    Average_Discount=('Discount','mean'),
    Average_Quantity=('Quantity','mean')
).sort_values('Total_Profit',ascending=False)

,Total_Sales,Total_Profit,Average_Discount,Average_Quantity
Product Name,,,,
Canon imageCLASS 2200 Advanced Copier,61599.824,25199.9280,0.120000,4.00
Fellowes PB500 Electric Punch Plastic Comb Binding Machine with Manual Bind,27453.384,7753.0390,0.240000,3.10
Hewlett Packard LaserJet 3310 Copier,18839.686,6983.8836,0.200000,4.75
Canon PC1060 Personal Laser Copier,11619.834,4570.9347,0.150000,4.75
"HP Designjet T520 Inkjet Large Format Printer - 24"" Color",18374.895,4094.9766,0.166667,4.00
...,...,...,...,...
GBC DocuBind P400 Electric Binding System,17965.068,-1878.1662,0.450000,4.50
Chromcraft Bull-Nose Wood Oval Conference Tables & Bases,9587.052,-2732.8608,0.250000,6.50
Cubify CubeX 3D Printer Triple Head Print,7999.980,-3839.9904,0.500000,4.00


Discovery :The Canon imageCLASS 2200 Advanced Copier generated the highest profit of all products, totaling an impressive 25,199.93	

The snippet below shows the general performance of the regions 

In [141]:
upper_outlier.groupby('Region').agg(
    Total_Sales=('Sales','sum'),
    Total_Profit=('Profit','sum'),
    Average_Discount=('Discount','mean'),
    Average_Quantity=('Quantity','mean')
).sort_values('Total_Profit',ascending=False)

,Total_Sales,Total_Profit,Average_Discount,Average_Quantity
Region,,,,
East,445893.7560,63497.0925,0.145509,5.206587
West,449647.5620,62532.9357,0.145321,5.195187
Central,317798.9774,36917.1810,0.162257,5.000000
South,264142.5095,24599.3859,0.125000,5.212871


The snippet below shows the performance based on Profit 

In [144]:
upper_outlier.groupby('Region')['Profit'].agg(
    Total_Profit='sum',
    Average_Profit='mean',
    Number_Of_Transactions='count'
).sort_values('Total_Profit',ascending=False)

,Total_Profit,Average_Profit,Number_Of_Transactions
Region,,,
East,63497.0925,190.111055,334
West,62532.9357,167.200363,374
Central,36917.1810,143.646619,257
South,24599.3859,121.779138,202


Discovery : The East Region generated the highest profit as compared to the other regions

Lets get back to business ...Our main goal is to try and understand why we have the variability right ?
So lets check the variability in terms of regions



NEXT QUESTION BECOMES WHICH REGION GENERATED THE HIGHEST PROFIT VARIATIONS OR BETTER STILL IN OTHER WORDS Are some regio responsible for the high variation in profit?


In [148]:
upper_outlier.groupby('Region')['Profit'].agg(['count','mean','std','min','max']).sort_values('std',ascending=False)

,count,mean,std,min,max
Region,,,,,
Central,257,143.646619,846.254081,-3701.8928,8399.9760
East,334,190.111055,704.623007,-6599.9780,5039.9856
South,202,121.779138,587.654593,-3839.9904,3177.4750
West,374,167.200363,478.327220,-3399.9800,6719.9808


In [ ]:
central=upper_outlier[upper_outlier['Region'] == 'Central']


In [ ]:
number_of_profitable_transactions=(central['Profit'] > 0).sum()
number_of_profitable_transactions

np.int64(167)

In [ ]:
number_of_loss_making_transactions=(central['Profit'] < 0).sum()
number_of_loss_making_transactions

np.int64(84)

In [159]:
profitable_transactions=central[central['Profit']> 0]

In [160]:
loss_making_transactions=central[central['Profit']< 0]

I Want to investigate what might be causing the losses in Central thus bringing about the high variability 


Which products are responsible for the largest total losses in Central



In [163]:
loss_making_transactions.groupby('Product Name')['Profit'].sum().sort_values()

Product Name
GBC DocuBind P400 Electric Binding System                                     -5552.8392
Fellowes PB500 Electric Punch Plastic Comb Binding Machine with Manual Bind   -3431.6730
Ibico EPK-21 Electric Binding System                                          -2929.4845
Lexmark MX611dhe Monochrome Laser Printer                                     -1869.9890
GBC DocuBind TL300 Electric Binding System                                    -1480.0335
                                                                                 ...    
Global Commerce Series Low-Back Swivel/Tilt Chairs                              -20.5584
Okidata C610n Printer                                                           -19.4700
Hon 4070 Series Pagoda Round Back Stacking Chairs                               -19.2588
Global Manager's Adjustable Task Chair, Storm                                   -12.0784
Fellowes Super Stor/Drawer Files                                                 -8.0775
Name: Pr

Discovery :Among the loss-making high-sales transactions in the Central region, the GBC DocuBind P400 Electric Binding System generated the largest total loss of approximately USD 5,552.84. It was followed by the Fellowes PB500 Electric Punch Plastic Comb Binding Machine with a loss of approximately KSh 3,431.67 and the Fellowes EPK-21 Electric Binding System with approximately USD 2,929.48 in losses.

This suggests that a small number of products are responsible for a substantial portion of the losses in Central's high-sales transactions, so these products should be investigated further to determine what is causing their negative profitability

In [169]:
loss_making_transactions.shape

(84, 21)

In [171]:
loss_making_transactions[['Sales','Profit','Discount','Quantity']].describe()

,Sales,Profit,Discount,Quantity
count,84.000000,84.000000,84.000000,84.000000
mean,1040.331302,-319.218487,0.354762,5.059524
std,985.192509,625.430004,0.150985,2.147294
min,509.488000,-3701.892800,0.200000,2.000000
25%,616.774800,-231.822150,0.300000,3.000000
50%,769.152500,-118.660650,0.300000,5.000000
75%,976.584000,-36.161700,0.320000,6.000000
max,8159.952000,-7.709400,0.800000,14.000000


In [172]:
loss_making_transactions['Profit'].sum()

np.float64(-26814.352899999998)

Discovery: The 84 loss-making high-sales transactions in Central generated a combined loss of approximately $26,814.35, with an average loss of $319.22 per transaction. Profit varies considerably among these transactions, with losses ranging from $7.71 to $3,701.89. The median loss of $118.66 is substantially smaller than the mean loss of $319.22, suggesting that a smaller number of transactions are responsible for disproportionately large losses

Earlier on we discovered that some products were generating the heavy losses but we did so on a surface level...Now I want to investigate it on a deeper level 

In [ ]:
product_losses=loss_making_transactions.groupby('Product Name').agg(
    Total_Sales=('Sales','sum'),
    Total_loss=('Profit',sum),
    Number_of_losses=('Profit','count'),
    Average_loss=('Profit','mean')
).sort_values('Total_loss')

,Total_Sales,Total_loss,Number_of_losses,Average_loss
Product Name,,,,
GBC DocuBind P400 Electric Binding System,3266.376,-5552.8392,2,-2776.4196
Fellowes PB500 Electric Punch Plastic Comb Binding Machine with Manual Bind,2287.782,-3431.6730,2,-1715.8365
Ibico EPK-21 Electric Binding System,1889.990,-2929.4845,1,-2929.4845
Lexmark MX611dhe Monochrome Laser Printer,11219.934,-1869.9890,2,-934.9945
GBC DocuBind TL300 Electric Binding System,896.990,-1480.0335,1,-1480.0335
...,...,...,...,...
Global Commerce Series Low-Back Swivel/Tilt Chairs,1439.088,-20.5584,2,-10.2792
Okidata C610n Printer,1362.900,-19.4700,1,-19.4700
Hon 4070 Series Pagoda Round Back Stacking Chairs,674.058,-19.2588,1,-19.2588


Lets calculate the percentage loss generated by each of the products above 